### PR TITLE
Introduce ClassTransformer

### PR DESCRIPTION
**Pull request by mkouba**
- that can be used to add a method to an existing class

This can be really useful when transforming existing classes in Quarkus.

I'm thinking of a follow-up PR that would allow users to add bytecode to an existing method as well...

### CODE REVIEW

**Comment by Ladicek** [comment]
I think this is a great idea, but I don't think we should expose standalone access to `MethodCreator`. How about an alternative to `ClassCreator` called, say, `ClassModifier`, that would expose access to other Gizmo APIs (like `MethodCreator` or possibly `FieldCreator` later?). I'm thinking something like this: https://github.com/Ladicek/gizmo/commit/628ddf516a344a18f61d8f579668628c0031ff9d

**Comment by mkouba** [comment]
> I think this is a great idea, but I don't think we should expose standalone access to `MethodCreator`. How about an alternative to `ClassCreator` called, say, `ClassModifier`, that would expose access to other Gizmo APIs (like `MethodCreator` or possibly `FieldCreator` later)?

Yes, that looks better. However, I don't think that `ClassModifier` is a good name because it reminds `Class.getModifiers()`... maybe `ClassTransformer`?

But I'm not quite sure we will be able to incorporate the "adding bytecode to an existing method" feature into this API...

Good point about the `FieldCreator`!

**Comment by Ladicek** [comment]
`ClassTransformer` is a good name, agree.

Do you have any idea how you wanted the "add bytecode to existing method" thing to look like? Do you want to add instructions at the beginning and/or the end of the method, or something more general? What I think we could expose relatively easily is to _rename_ a method -- then you can create a new one that would replace the original and you call to it whenever you need. WDYT?

**Comment by mkouba** [comment]
> `ClassTransformer` is a good name, agree.

:+1: 

> 
> Do you have any idea how you wanted the "add bytecode to existing method" thing to look like? Do you want to add instructions at the beginning and/or the end of the method, or something more general? What I think we could expose relatively easily is to _rename_ a method -- then you can create a new one that would replace the original and you call to it whenever you need. WDYT?

So I was thinking of adding some bytecode at the beginning/end of a method. For example if you need to initialize the fields in the constructor.

Renaming a method would be definitely a useful addition. The same applies to modyfing the modifiers ;-)

File: src/test/java/io/quarkus/gizmo/ClassTransformerTest.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        classTransformer.close();
+
+        cv.visitEnd();
+        cl.write(className, cv.toByteArray());
+
+        Class<?> clazz = cl.loadClass("com.MyTest");
+        assertTrue(clazz.isSynthetic());
+        MyInterface myInterface = (MyInterface) clazz.getDeclaredConstructor().newInstance();
+        Method transformMethod = clazz.getMethod("transform", String.class);
+        assertTrue(Modifier.isFinal(transformMethod.getModifiers()));
```
Comment (Ladicek) [inline src/test/java/io/quarkus/gizmo/ClassTransformerTest.java:97]: I would use try-with-resources for the `ClassTransformer` and use reflection here to also assert something about the field.

**Review by Ladicek — APPROVED** [review]
Otherwise LGTM.

File: src/test/java/io/quarkus/gizmo/ClassTransformerTest.java (hunk 1 above)
Comment (mkouba) [reply src/test/java/io/quarkus/gizmo/ClassTransformerTest.java:100]: I don't think that we need to test the `AutoCloseable` functionality... Field assertion added.

**Comment by mkouba** [comment]
> An interesting thing from API design perspective is that modifying existing stuff requires more careful handling of the `ClassVisitor`. Adding new things can be pretty much willy nilly (until `visitEnd()` is called), but modifying existing requires the `ClassTransformer` to be notified of all events that happen on the visitor.
> 
> So either the `ClassTransformer` will itself be a `ClassVisitor`, or it will expose one, something like that.

Hm, sounds like fun ;-). Anyway, let's add one thing at a time.

**Comment by Ladicek** [comment]
Ah, there's one other thing worth discussing: both `FieldDescriptor` and `MethodDescriptor` require that the user knows the full [binary] name of the class they're transforming, even though that typically is not necessary (e.g. `FieldCreatorImpl` never uses that part of the `FieldDescriptor`, and I suspect `MethodCreatorImpl` is the same).

Maybe we should expose `addField` / `addMethod` methods that don't require the full descriptor?

**Comment by mkouba** [comment]
> Ah, there's one other thing worth discussing: both `FieldDescriptor` and `MethodDescriptor` require that the user knows the full [binary] name of the class they're transforming, even though that typically is not necessary (e.g. `FieldCreatorImpl` never uses that part of the `FieldDescriptor`, and I suspect `MethodCreatorImpl` is the same).
> 
> Maybe we should expose `addField` / `addMethod` methods that don't require the full descriptor?

Hm, I used `FieldDescriptor` and `MethodDescriptor` because they declare a lot of useful factory methods, that also handle the conversion to internal class names, etc. And I'd like to avoid adding many `addMethod()`/`addField()` variants to the `ClassTransformer`.

**Comment by Ladicek** [comment]
Mirroring what `ClassCreator` does would be fine IMHO. Perhaps even expected :shrug:

**Comment by mkouba** [comment]
> Mirroring what `ClassCreator` does would be fine IMHO. Perhaps even expected shrug

The thing is that `ClassCreator` knows the `className` but the `ClassVisitor` does not, so we couldn't really use the `MethodDescriptor` internally :disappointed: .

**Comment by Ladicek** [comment]
Just use `null` as the class name? Should work, and if there's a NPE somewhere, it's a bug.

We could also change `ClassTransformer` constructor to take the name of the transformed class.

**Comment by mkouba** [comment]
> Just use `null` as the class name? Should work, and if there's a NPE somewhere, it's a bug.

I think that `DescriptorUtils.objectToInternalClassName()` would fail with IAE... why do you think it's a bug? I would say it's an incomplete descriptor definition instead ;-).

> We could also change `ClassTransformer` constructor to take the name of the transformed class.

That would work.

**Comment by Ladicek** [comment]
Fair enough, it's probably better if we don't allow constructing incomplete descriptors :-)

**Comment by mkouba** [comment]
I think that this PR is now ready! ;-)

**Review by Ladicek — APPROVED** [review]

**Comment by Ladicek** [comment]
Hi @mkouba, I know I promised you to merge this PR, but then I started thinking and realized that the API we have now doesn't allow more complex transformations that we talked about above. I came up with an API that I believe is better (and allowed implementing method rename fairly easily as a proof): https://github.com/Ladicek/gizmo/commits/class-transformer Please take a look when you have time, and thanks for understanding :-)

**Comment by mkouba** [comment]
> Hi @mkouba, I know I promised you to merge this PR, but then I started thinking and realized that the API we have now doesn't allow more complex transformations that we talked about above. I came up with an API that I believe is better (and allowed implementing method rename fairly easily as a proof): https://github.com/Ladicek/gizmo/commits/class-transformer Please take a look when you have time, and thanks for understanding :-)

It looks good. But I'm not so sure about the name of the `ClassTransformer#wrap(ClassVisitor delegate)` method. It's not very informative.

**Comment by Ladicek** [comment]
Fair enough, WDYT about `wrapVisitor`? Or `apply[ToVisitor]`?

**Comment by mkouba** [comment]
> Fair enough, WDYT about `wrapVisitor`? Or `apply[ToVisitor]`?

`apply()`/`applyTo()` sounds good.

**Comment by Ladicek** [comment]
Superseded by #158.

**Pull request closed without merging**